### PR TITLE
[SPARK-39537][SQL] Ensure that the expressions order returned by `InferFiltersFromConstraints`

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, `InferFiltersFromConstraints` construct a new `Filter` which condition without the same order in scala 2.12 and 2.13

It seems `--` of `ExpressionSet` cannot guarantee the order of elements.

The behavior lead to https://github.com/apache/spark/pull/35947#discussion_r901193520

### Why are the changes needed?
Fix bug

### Does this PR introduce _any_ user-facing change?
'Yes'.
The behavior will be consistent between scala 2.12 and 2.13

### How was this patch tested?
Exists tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1272,8 +1272,11 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan]
   private def inferFilters(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
     _.containsAnyPattern(FILTER, JOIN)) {
     case filter @ Filter(condition, child) =>
-      val newFilters = filter.constraints --
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: is it possible to override `--` in `ExpressionSet`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: cc @LuciferYang

**Comment by beliefer** [comment]
ping @HyukjinKwon @huaxingao cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: Yes, after SPARK-39520, this issue no longer exists， and @HyukjinKwon already mark SPARK-39523 as RESOLVED

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: It seems https://github.com/apache/spark/pull/36925 fixed this issue ? @LuciferYang

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: [SPARK-39520](https://issues.apache.org/jira/browse/SPARK-39520) was not merged to 3.3, do we have this issue in 3.3?

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: > It seems #36925 fixed this issue ? @LuciferYang

Yes

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: > [SPARK-39520](https://issues.apache.org/jira/browse/SPARK-39520) was not merged to 3.3, do we have this issue in 3.3?

Let me double check this

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: @cloud-fan @LuciferYang I have tested and the issue no longer exists. I will close this PR.

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1275]: > [SPARK-39520](https://issues.apache.org/jira/browse/SPARK-39520) was not merged to 3.3, do we have this issue in 3.3?

@cloud-fan 3.3 no this issue due to SPARK-38836 not in branch-3.3, the `--` method hasn't been deleted, and I manually check the relevant suites(inlcude TPCDSV1_4_PlanStabilitySuite, AdaptiveQueryExecSuite, ExpressionSetSuite, JDBCV2Suite, GeneratorFunctionSuite), all passed.

**Pull request closed without merging**